design/worker-config-drive.md

```diff
@@ -0,0 +1,196 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# worker-config-drive
+
+## Status
+
+provisional
+
+## Table of Contents
+
+<!--ts-->
+   * [worker-config-drive](#worker-config-drive)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: dhellmann, at: 2019-03-14T09:25-0400 -->
+
+<!--te-->
+
+## Summary
+
+Provisioning hosts requires two separate images. The first is the
+primary target image for the host, and contains the operating system
+and other software that the host will run. These are generally
+reusable across many hosts. The second image is the "config drive
+image", which contains configuration settings passed to the target
+image by writing an ISO to a separate partition accessible when the
+host boots. The config drive image is often, but not always,
+customized for each host. It is always often customized for the role a
+host plays (master, worker, etc.).
+
+In order to bring a host into the cluster as a worker we have to give
+it the correct Ignition configuration as it boots.  The target images
+we are using for nodes in an OpenShift (and eventually Kubernetes)
+cluster are RHEL CoreOS images with Ignition configured to look for
+its data in `/openstack/latest` in the config drive. This setting is
+built in to the image, and is not currently configurable.
+
+The Ignition file contents are stored in a Secret within the
+kubernetes database because they contain certificate authority
+information. There are different Ignition settings for worker and
+master nodes, so something will need to know the intended role of each
+host being provisioned in order to select the correct settings.
+
+Something needs to build a config drive image containing the ignition
+file and the baremetal operator needs to pass that formatted image to
+Ironic as part of provisioning the host.
```

Comment:
Why not make Ironic build it for us?

Reply:
How would we tell ironic where to put the ignition config file?

Reply:
From what I learned, it's passed as `user_data`, which ironic supports.

Reply:
What format does that field take? Doesn't something need to tell Ironic to put the file in the right location within the config drive?

Reply:
So, you call ironic with something like `"configdrive": {"user_data": "<some text>"}`. The `<some text>` then ends up written to the generated configdrive at `openstack/latest/user_data` as text. IIUC that's exactly what Ignition needs.

Reply:
Gophercloud does the same.  As of now, I have seen nothing that expects, creates, or consumes user data from anything other than `openstack/latest/user_data`, with the exception that it's possible to make `latest` a version instead.